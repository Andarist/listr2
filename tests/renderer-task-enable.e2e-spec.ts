/* eslint-disable @typescript-eslint/no-empty-function */
import { mockProcessExit, mockProcessStderr, mockProcessStdout } from 'jest-mock-process'

import { Listr } from '@root'

describe('show task enable', () => {
  let mockExit: jest.SpyInstance<never, [number?]>

  // eslint-disable-next-line @typescript-eslint/ban-types
  let mockStdout: jest.SpyInstance<boolean, [string, string?, Function?]>

  // eslint-disable-next-line @typescript-eslint/ban-types
  let mockStderr: jest.SpyInstance<boolean, [string, string?, Function?]>

  process.stdout.isTTY = true

  beforeEach(async () => {
    mockExit = mockProcessExit()
    mockStdout = mockProcessStdout()
    mockStderr = mockProcessStderr()
  })

  afterEach(async () => {
    mockExit.mockRestore()
    mockStdout.mockRestore()
    mockStderr.mockRestore()
    jest.clearAllMocks()
  })

  // JeLYvYj1f4ddSjgwkUp7Nr8hICHLdEsU
  it('should enable task with general context', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: (ctx): void => {
            ctx.skip = true
          }
        },

        {
          title: 'This task will never execute.',
          enabled: (ctx): boolean => !ctx.skip,
          task: (): void => {}
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expect(mockStdout.mock.calls).toMatchSnapshot('JeLYvYj1f4ddSjgwkUp7Nr8hICHLdEsU-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('JeLYvYj1f4ddSjgwkUp7Nr8hICHLdEsU-err')
    expect(mockExit.mock.calls).toMatchSnapshot('JeLYvYj1f4ddSjgwkUp7Nr8hICHLdEsU-exit')
  })

  // 2PeJ1GS5SrNdC7DXMv8nHAQNfire6UYB
  it('should enable task with general context in subtasks', async () => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: (ctx): void => {
            ctx.skip = true
          }
        },

        {
          title: 'This task will show subtasks.',
          task: (_, task): Listr =>
            task.newListr(
              [
                {
                  title: 'This task will execute.',
                  task: (): void => {}
                },

                {
                  title: 'This task will never execute.',
                  enabled: (ctx): boolean => !ctx.skip,
                  task: (): void => {}
                }
              ],
              { rendererOptions: { collapse: false }, concurrent: true }
            )
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expect(mockStdout.mock.calls).toMatchSnapshot('2PeJ1GS5SrNdC7DXMv8nHAQNfire6UYB-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('2PeJ1GS5SrNdC7DXMv8nHAQNfire6UYB-err')
    expect(mockExit.mock.calls).toMatchSnapshot('2PeJ1GS5SrNdC7DXMv8nHAQNfire6UYB-exit')
  })
})
