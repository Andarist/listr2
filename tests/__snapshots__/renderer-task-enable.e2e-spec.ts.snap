// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show task enable should enable task with general context in subtasks: 2PeJ1GS5SrNdC7DXMv8nHAQNfire6UYB-err 1`] = `[]`;

exports[`show task enable should enable task with general context in subtasks: 2PeJ1GS5SrNdC7DXMv8nHAQNfire6UYB-exit 1`] = `[]`;

exports[`show task enable should enable task with general context in subtasks: 2PeJ1GS5SrNdC7DXMv8nHAQNfire6UYB-out 1`] = `
[
  [
    "‚óº This task will execute.
",
  ],
  [
    "[2K[1A[2K[G‚óº This task will execute.
‚óº This task will show subtasks.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
‚óº This task will show subtasks.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚óº This task will show subtasks.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚ùØ This task will show subtasks.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚ùØ This task will show subtasks.
  ‚óº This task will execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚ùØ This task will show subtasks.
  ‚ùØ This task will execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚ùØ This task will show subtasks.
  ‚úî This task will execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚úî This task will show subtasks.
  ‚úî This task will execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[1A[2K[G",
  ],
  [
    "‚úî This task will execute.
‚úî This task will show subtasks.
  ‚úî This task will execute.
",
  ],
]
`;

exports[`show task enable should enable task with general context: JeLYvYj1f4ddSjgwkUp7Nr8hICHLdEsU-err 1`] = `[]`;

exports[`show task enable should enable task with general context: JeLYvYj1f4ddSjgwkUp7Nr8hICHLdEsU-exit 1`] = `[]`;

exports[`show task enable should enable task with general context: JeLYvYj1f4ddSjgwkUp7Nr8hICHLdEsU-out 1`] = `
[
  [
    "‚óº This task will execute.
",
  ],
  [
    "[2K[1A[2K[G‚óº This task will execute.
‚óº This task will never execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
‚óº This task will never execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚óº This task will never execute.
",
  ],
  [
    "[2K[1A[2K[1A[2K[G‚úî This task will execute.
",
  ],
  [
    "[2K[1A[2K[G",
  ],
  [
    "‚úî This task will execute.
",
  ],
]
`;
